csrc/multidevice/device_mesh.cpp

```diff
@@ -56,4 +106,69 @@ int64_t DeviceMesh::size(const ParallelType parallel_type) const {
   return size();
 }
 
+std::vector<int64_t> DeviceMesh::getIndices(const DeviceIdxType device) const {
+  auto global_idx = idxOf(device);
+  if (global_idx == -1) {
+    return {};
+  }
+  std::vector<int64_t> indices(shape_.size());
+  int64_t accumulated_size = 1;
+  for (int64_t i = (int64_t)shape_.size() - 1; i >= 0; i--) {
+    indices[i] = (global_idx / accumulated_size) % shape_[i];
+    accumulated_size *= shape_[i];
+  }
+  return indices;
+}
+
+DeviceIdxType DeviceMesh::maxDeviceId() const {
+  return *std::max_element(vector_.begin(), vector_.end());
+}
+
+namespace {
+size_t ptypeToAxis(ParallelType ptype, size_t naxis) {
+  NVF_ERROR(
+      isParallelTypeDeviceDim(ptype),
+      "Attempting to index into DeviceMesh with a non-device parallel type",
+      ptype);
+  size_t offset =
+      static_cast<size_t>(ptype) - static_cast<size_t>(ParallelType::DIDx);
+
+  NVF_ERROR(
+      offset < naxis,
+      "DeviceMesh has ",
+      naxis,
+      " dimensions, but requesting slice for ",
+      ptype);
+  return naxis - 1 - offset;
+}
+} // namespace
+
+std::vector<DeviceIdxType> DeviceMesh::getSlice(
+    DeviceIdxType deviceId,
+    ParallelType ptype) const {
+  size_t axis = ptypeToAxis(ptype, shape_.size());
+  auto indices = getIndices(deviceId);
+  NVF_ERROR(
+      !indices.empty(), "Device ", deviceId, " is not in DeviceMesh ", vector_);
+
+  int64_t offset = 0;
+  int64_t stride = 1;
+  int64_t accumulated_size = 1;
+  for (size_t i = shape_.size() - 1; i >= 0; i--) {
```

Comment:
```suggestion
  for (auto i = static_cast<int64_t>(shape_.size()) - 1; i >= 0; i--) {
```
I'm surprised the current code doesn't run into a infinite loop. `i` is size_t so always >=0. (That's another reason for https://google.github.io/styleguide/cppguide.html#Integer_Types). 

I also recommend a DeviceMesh::rank() that returns shape_.size() as an int64_t. This will potentially save us many static_casts elsewhere.

Reply:
Sounds good for the int64_t size(). Hopefully it can clean up some casts, and save any issues with size_t wrap around..which did happen before..